### DISCUSSION TITLE
Module 'Doc' import error when using FastAPI

### DISCUSSION THREAD

**Discussion by kojomensahonums** [fastapi/fastapi#10476] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
%%capture
# Install required libraries for web api
#!pip -q install fastapi
!pip -q install pyngrok
!pip -q install uvicorn
!pip -q install nest_asyncio
!pip -q install python-multipart
!pip uninstall fastapi typing-extensions -y
!pip install fastapi typing-extensions==4.8.0 --no-cache-dir

-----------------------------------------------------------------------------
# Import libraries
import uvicorn
from typing_extensions import Annotated#, Doc
from fastapi import FastAPI
from fastapi.middleware.cors import CORSMiddleware
from pyngrok import ngrok
from pydantic import BaseModel
import nest_asyncio
import shutil

# Import models for serialisation/ deserialisation
from pydantic import BaseModel
import base64
import io
import wave

nest_asyncio.apply()

app = FastAPI()

app.add_middleware(
    CORSMiddleware,
    allow_origins=['*'],
    allow_credentials=True,
    allow_methods=['*'],
    allow_headers=['*'],
)
```

### Description

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
[<ipython-input-10-86c7111d233f>](https://localhost:8080/#) in <cell line: 4>()
      2 import uvicorn
      3 from typing_extensions import Annotated#, Doc
----> 4 from fastapi import FastAPI
      5 from fastapi.middleware.cors import CORSMiddleware
      6 from pyngrok import ngrok

6 frames
[/usr/local/lib/python3.10/dist-packages/fastapi/__init__.py](https://localhost:8080/#) in <module>
      5 from starlette import status as status
      6 
----> 7 from .applications import FastAPI as FastAPI
      8 from .background import BackgroundTasks as BackgroundTasks
      9 from .datastructures import UploadFile as UploadFile

[/usr/local/lib/python3.10/dist-packages/fastapi/applications.py](https://localhost:8080/#) in <module>
     14 )
     15 
---> 16 from fastapi import routing
     17 from fastapi.datastructures import Default, DefaultPlaceholder
     18 from fastapi.exception_handlers import (

[/usr/local/lib/python3.10/dist-packages/fastapi/routing.py](https://localhost:8080/#) in <module>
     20 )
     21 
---> 22 from fastapi import params
     23 from fastapi._compat import (
     24     ModelField,

[/usr/local/lib/python3.10/dist-packages/fastapi/params.py](https://localhost:8080/#) in <module>
      3 from typing import Any, Callable, Dict, List, Optional, Sequence, Union
      4 
----> 5 from fastapi.openapi.models import Example
      6 from pydantic.fields import FieldInfo
      7 from typing_extensions import Annotated, deprecated

[/usr/local/lib/python3.10/dist-packages/fastapi/openapi/models.py](https://localhost:8080/#) in <module>
      2 from typing import Any, Callable, Dict, Iterable, List, Optional, Set, Type, Union
      3 
----> 4 from fastapi._compat import (
      5     PYDANTIC_V2,
      6     CoreSchema,

[/usr/local/lib/python3.10/dist-packages/fastapi/_compat.py](https://localhost:8080/#) in <module>
     18 )
     19 
---> 20 from fastapi.exceptions import RequestErrorModel
     21 from fastapi.types import IncEx, ModelNameMap, UnionType
     22 from pydantic import BaseModel, create_model

[/usr/local/lib/python3.10/dist-packages/fastapi/exceptions.py](https://localhost:8080/#) in <module>
      4 from starlette.exceptions import HTTPException as StarletteHTTPException
      5 from starlette.exceptions import WebSocketException as StarletteWebSocketException
----> 6 from typing_extensions import Annotated, Doc  # type: ignore [attr-defined]
      7 
      8 

ImportError: cannot import name 'Doc' from 'typing_extensions' (/usr/local/lib/python3.10/dist-packages/typing_extensions.py)

---------------------------------------------------------------------------
NOTE: If your import is failing due to a missing package, you can
manually install dependencies using either !pip or !apt.

To view examples of installing some common dependencies, click the
"Open Examples" button below.
------------------------------------

### Operating System

Windows

### FastAPI Version

0.104.0

### Pydantic Version

1.10.13

### Python Version

3.10

### Additional Context

I am running in Google Colab and would like to continue in that manner. I tried the suggestion to uninstall and install FastAPI without cache so I could install the newer version of typing-extensions but it did not work.

**Comment by tiangolo** [comment, marked as answer]
It seems in some way you ended up with a version of typing-extensions lower that 4.8.0, even though that's what you have in the example cell.

Please try following the normal approach, with local Python, a local virtual environment, without any additional factors, and see if that works. If it works, then the problem is in your Colab environment and related to that.

I'm also not sure how would you work with FastAPI in Colab, but anyway, that's another discussion, just try that first. 🤓

**Reply by kojomensahonums** [reply]
The reason for using Colab is that I have ML models running in there. I'm trying to create API calls to the models running on Colab and that's why I want it working there. The approach you suggested works but I still need a fix for Google Colab (has GPU accessibility).

**Comment by mcaferulker** [comment]
I've got exact same issue. Again, in Colab, trying to run a ML model. I've tried lower versions but it didn't work for me.

**Comment by MatrixManAtYrService** [comment]
I ran across this error because I was doing dumb things with multiple package managers.  I had a `flake.nix` which was placing the following string in my PATH: `/nix/store/...typing-extensions-4.7.1/lib/python3.11/site-packages/typing_extensions.py`. Then I used Poetry to install `fastapi = "^0.104.0"`.  FastAPI depends on typing-extensions 4.8.0, which has `Doc` defined.

I bet that poetry did install the correct version, but for some reason the nix-supplied (older) version was coming first in my PATH, so that version (which does not have `Doc` defined) was being used instead of the correct one.

Probably something similar is happening with Colab.  You've got a new version of FastAPI, but something else depends on a too-old version of `typing_extensions` and for some (broken) reason, the older version of the dependency is taking precedence.   

I don't know Colab myself, but maybe you can get the stuff you care about in a virtual environment so that it's isolated from whatever else is in the environment?

If not, give google hell for:
1. letting their dependencies mingle with yours *at all*
2. putting their dependencies before yours on the PATH as it appears to your code

**Comment by kojomensahonums** [comment]
@MatrixManAtYrService I tried with a virtual environment in Colab and still got thrown the same error. Maybe running the virtual environment in Colab was a bad idea to start with, regarding this issue.

**Comment by relsunkaev** [comment]
I'm currently having the same issue when using datadog python layers with [`serverless-plugin-datadog`](https://github.com/DataDog/serverless-plugin-datadog). Works perfectly fine locally and without the layers. I know it's probably an issue on their end but to anyone looking you'll have to downgrade FastAPI for it to work.

**Comment by reisman234** [comment]
I do now see also this error in the python log when does the test discovery. This ImportError does not happen when I open the python CLI and import it manually. 

My environment is a dev container in vscode with python3.10

```
2023-11-23 10:58:47.646 [error] Unittest test discovery error 
 Failed to import test module: test.test_service_api
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/unittest/loader.py", line 436, in _find_test_path
    module = self._get_module_from_name(name)
  File "/usr/local/lib/python3.10/unittest/loader.py", line 377, in _get_module_from_name
    __import__(name)
  File "/workspaces/project/test/test_service_api.py", line 5, in <module>
    from fastapi.testclient import TestClient
  File "/usr/local/lib/python3.10/site-packages/fastapi/__init__.py", line 7, in <module>
    from .applications import FastAPI as FastAPI
  File "/usr/local/lib/python3.10/site-packages/fastapi/applications.py", line 16, in <module>
    from fastapi import routing
  File "/usr/local/lib/python3.10/site-packages/fastapi/routing.py", line 22, in <module>
    from fastapi import params
  File "/usr/local/lib/python3.10/site-packages/fastapi/params.py", line 5, in <module>
    from fastapi.openapi.models import Example
  File "/usr/local/lib/python3.10/site-packages/fastapi/openapi/models.py", line 4, in <module>
    from fastapi._compat import (
  File "/usr/local/lib/python3.10/site-packages/fastapi/_compat.py", line 20, in <module>
    from fastapi.exceptions import RequestErrorModel
  File "/usr/local/lib/python3.10/site-packages/fastapi/exceptions.py", line 6, in <module>
    from typing_extensions import Annotated, Doc  # type: ignore [attr-defined]
ImportError: cannot import name 'Doc' from 'typing_extensions' (/home/vscode/.vscode-server/extensions/ms-python.python-2023.20.0/pythonFiles/lib/python/typing_extensions.py)
```

**Comment by yangcheng** [comment]
i have a similar error with vscode and python 3.12

**Comment by reisman234** [comment]
I did a small setup with fastapi, uvicorn and an example application including a test case, also from within a dev container but this time the dev container uses a python3.11 environment. The Sources can be found in this [repo](https://github.com/reisman234/fastapi-demo)

At first, I used pytest as a testing framework, which worked as expected and every test ran successfully. After a change in the "unittest" framework, the same error occurred as before.

```
2023-11-24 11:35:26.273 [error] Unittest test discovery error 
 Failed to import test module: test_app
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/unittest/loader.py", line 419, in _find_test_path
    module = self._get_module_from_name(name)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/unittest/loader.py", line 362, in _get_module_from_name
    __import__(name)
  File "/workspaces/dev-container/test/test_app.py", line 3, in <module>
    from fastapi.testclient import TestClient
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/__init__.py", line 7, in <module>
    from .applications import FastAPI as FastAPI
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/applications.py", line 16, in <module>
    from fastapi import routing
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/routing.py", line 22, in <module>
    from fastapi import params
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/params.py", line 5, in <module>
    from fastapi.openapi.models import Example
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/openapi/models.py", line 4, in <module>
    from fastapi._compat import (
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/_compat.py", line 20, in <module>
    from fastapi.exceptions import RequestErrorModel
  File "/home/vscode/.local/lib/python3.11/site-packages/fastapi/exceptions.py", line 6, in <module>
    from typing_extensions import Annotated, Doc  # type: ignore [attr-defined]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: cannot import name 'Doc' from 'typing_extensions' (/home/vscode/.vscode-server/extensions/ms-python.python-2023.20.0/pythonFiles/lib/python/typing_extensions.py)
2023-11-24 11:35:26.273 [info] ResultResolver EOT received for discovery.
```

**Reply by reisman234** [reply]
I switched back to `fastapi==0.103.2`, with this version, the test discovery did work again.

**Reply by shdw2x** [reply]
In my case, I was getting the same error and I tried to change the Python extension's version in VS Code. Upgrading it to the latest version, then restarting VS Code, and then switching the Python extension's version to pre-release solved the problem for me. Thank you for the suggestion @free-bit

![image](https://github.com/tiangolo/fastapi/assets/6889642/2f633d2c-ebdc-4624-b315-5fd7fc7a322f)

**Reply by mertozlutiras** [reply]
downgrading to `fastapi==0.103.2` worked for me.

**Comment by ArturFormella** [comment]
I have the same problem:

```
(.venv) (base) ➜  server git:(main) ✗ pip install fastapi==0.104.1 https://github.com/python/typing_extensions/archive/main.zip
Collecting https://github.com/python/typing_extensions/archive/main.zip
  Using cached https://github.com/python/typing_extensions/archive/main.zip
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting fastapi==0.104.1
  Obtaining dependency information for fastapi==0.104.1 from https://files.pythonhosted.org/packages/f3/4f/0ce34195b63240b6693086496c9bab4ef23999112184399a3e88854c7674/fastapi-0.104.1-py3-none-any.whl.metadata
  Using cached fastapi-0.104.1-py3-none-any.whl.metadata (24 kB)
Requirement already satisfied: anyio<4.0.0,>=3.7.1 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from fastapi==0.104.1) (3.7.1)
Requirement already satisfied: pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,!=2.1.0,<3.0.0,>=1.7.4 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from fastapi==0.104.1) (2.5.2)
Requirement already satisfied: starlette<0.28.0,>=0.27.0 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from fastapi==0.104.1) (0.27.0)
Requirement already satisfied: idna>=2.8 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from anyio<4.0.0,>=3.7.1->fastapi==0.104.1) (3.5)
Requirement already satisfied: sniffio>=1.1 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from anyio<4.0.0,>=3.7.1->fastapi==0.104.1) (1.3.0)
Requirement already satisfied: annotated-types>=0.4.0 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,!=2.1.0,<3.0.0,>=1.7.4->fastapi==0.104.1) (0.6.0)
Requirement already satisfied: pydantic-core==2.14.5 in /druga/xtts-streaming-server/.venv/lib/python3.11/site-packages (from pydantic!=1.8,!=1.8.1,!=2.0.0,!=2.0.1,!=2.1.0,<3.0.0,>=1.7.4->fastapi==0.104.1) (2.14.5)
Using cached fastapi-0.104.1-py3-none-any.whl (92 kB)
Installing collected packages: fastapi
  Attempting uninstall: fastapi
    Found existing installation: fastapi 0.104.0
    Uninstalling fastapi-0.104.0:
      Successfully uninstalled fastapi-0.104.0
Successfully installed fastapi-0.104.1
```
uvicorn "main:app" --host 0.0.0.0 --port 80
```

Traceback (most recent call last):
  File "/home/artur/anaconda3/bin/uvicorn", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/site-packages/uvicorn/main.py", line 416, in main
    run(
  File "/home/artur/anaconda3/lib/python3.11/site-packages/uvicorn/main.py", line 587, in run
    server.run()
  File "/home/artur/anaconda3/lib/python3.11/site-packages/uvicorn/server.py", line 61, in run
    return asyncio.run(self.serve(sockets=sockets))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/asyncio/runners.py", line 190, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "uvloop/loop.pyx", line 1517, in uvloop.loop.Loop.run_until_complete
  File "/home/artur/anaconda3/lib/python3.11/site-packages/uvicorn/server.py", line 68, in serve
    config.load()
  File "/home/artur/anaconda3/lib/python3.11/site-packages/uvicorn/config.py", line 467, in load
    self.loaded_app = import_from_string(self.app)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/site-packages/uvicorn/importer.py", line 21, in import_from_string
    module = importlib.import_module(module_str)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/artur/anaconda3/lib/python3.11/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "<frozen importlib._bootstrap>", line 1204, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 940, in exec_module
  File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
  File "/druga/xtts-streaming-server/server/main.py", line 3, in <module>
    from fastapi import (
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/__init__.py", line 7, in <module>
    from .applications import FastAPI as FastAPI
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/applications.py", line 16, in <module>
    from fastapi import routing
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/routing.py", line 22, in <module>
    from fastapi import params
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/params.py", line 5, in <module>
    from fastapi.openapi.models import Example
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/openapi/models.py", line 4, in <module>
    from fastapi._compat import (
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/_compat.py", line 20, in <module>
    from fastapi.exceptions import RequestErrorModel
  File "/home/artur/anaconda3/lib/python3.11/site-packages/fastapi/exceptions.py", line 6, in <module>
    from typing_extensions import Annotated, Doc  # type: ignore [attr-defined]
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ImportError: cannot import name 'Doc' from 'typing_extensions' (/home/artur/anaconda3/lib/python3.11/site-packages/typing_extensions.py)

```

**Comment by MatrixManAtYrService** [comment]
It took [some hunting](https://discourse.nixos.org/t/help-using-poetry-in-a-flake-devshell/36874/2?u=mattrixman) but in my case it turned out that both `mypy` and `psycopg` were depending on typing_extensions at v2.7.1, which is too old for new versions of fastapi.  I would've expected my virtual environment to shadow these dependencies (which were in the outer environment) with the correct versions, but I guess that's not how it works?

**Comment by boltomli** [comment]
Due to requirement from odigos, I ended up with fastapi < 0.104 and typing_extensions < 4.8

**Comment by brainfo** [comment]
Hi, I encountered the same error on google colab.
Anyone fixed this?
